### DISCUSSION TITLE
Support multiple types of ABI format in useContract calls

### DISCUSSION THREAD

**Discussion by MidnightLightning** [wevm/wagmi#1763] [Ideas]
The [`useContract`](https://wagmi.sh/react/hooks/useContract) hook (and all hooks with that prefix) take an `abi` parameter that has to be in the Solidity language's defined [JSON-format](https://docs.soliditylang.org/en/v0.8.15/abi-spec.html#json).

However, the output of that contract is an [Ethers Contract](https://docs.ethers.org/v5/api/contract/contract/), and that library supports creating Contracts using ABIs in [multiple formats](https://docs.ethers.org/v5/api/utils/abi/formats/#abi-formats). Notably, they have a "Human-Readable" format that is much more concise than the Solidity JSON format, allowing for much more readable code.

The Ethers library has a `ContractInterface` type that is what it defines as needed for input into the Contract constructor. Can the wagmi hooks be updated to accept any form of `ethers.ContractInterface`? And if the hooks internally need to parse which functions are doing what, internally it could use the Ethers library utilities to [convert between the formats](https://docs.ethers.org/v5/api/utils/abi/formats/#abi-formats--converting-between-formats)?

**Comment by tmm** [comment]
Thanks for the feedback!

It's possible that wagmi supports some sort of human-readable ABI format in the future, but for now, we are going to continue only supporting the ABI specification as it's the only concrete standard and allows for [inferring TypeScript types](https://wagmi.sh/react/typescript) directly from ABIs.

---

Couple notes:

- I would consider ethers usage in wagmi to be an implementation detail so the fact they it supports non-ABI spec formats isn't relevant at the moment.
- You can use [`@wagmi/cli`](https://wagmi.sh/cli/getting-started) to manage ABIs automatically (fetching from block explorers, Foundry/Hardhat projects, etc.) and generate React Hooks.

**Comment by MidnightLightning** [comment]
I would push back a little on ethers library usage being an "implementation detail" because the `useContract` hook returns that object to the consuming application. Hence the consuming applications need to have that library as a dependency, and know the capabilities of that library's Contract class, to use that return value effectively. If wagmi were to change which library it used for that Contract representation, that would require a major change on all consuming client applications.

Currently I am doing the conversion before using the wagmi hook:

```js
const abi = new ethers.utils.Interface([
  'function foo() external view returns (uint256)',
  'function bar(address) external view returns (address)',
])
const myContract = useContract({
  address: '0x...',
  abi: JSON.parse(abi.format(ethers.utils.FormatTypes.json) as string),
})
```

But if wagmi did that conversion internally, it would still have access to the JSON-formatted ABI for type-checking (that benefit wouldn't be lost, I believe?), and reduce the boilerplate code needed on the client side for simple contract interfaces.

**Comment by MidnightLightning** [comment]
Seems this feature might have been implemented recently?

https://twitter.com/wagmi_sh/status/1631017966505541633?t=lCfvO49gytWreryxleaeIA&s=19

What version is this new ability added into?

**Reply by tmm** [reply]
It a [ABIType](https://abitype.dev/) feature and doesn't exist in wagmi quite yet.
